pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java

```diff
@@ -458,13 +460,42 @@ public boolean routingExists(String tableNameWithType) {
    * <p>NOTE: The broker request should already have the table suffix (_OFFLINE or _REALTIME) appended.
    */
   @Nullable
+  @Override
   public RoutingTable getRoutingTable(BrokerRequest brokerRequest) {
     String tableNameWithType = brokerRequest.getQuerySource().getTableName();
     RoutingEntry routingEntry = _routingEntryMap.get(tableNameWithType);
     if (routingEntry == null) {
       return null;
     }
     InstanceSelector.SelectionResult selectionResult = routingEntry.calculateRouting(brokerRequest);
+    return constructRoutingTableFromInstanceSelectionResult(tableNameWithType, selectionResult);
+  }
+
+  @Override
+  public RoutingTable getRoutingTable(String tableNameWithType) {
+    RoutingEntry routingEntry = _routingEntryMap.get(tableNameWithType);
+    if (routingEntry == null) {
+      return null;
+    }
+    InstanceSelector.SelectionResult selectionResult = routingEntry.calculateRouting(tableNameWithType);
+    return constructRoutingTableFromInstanceSelectionResult(tableNameWithType, selectionResult);
+  }
+
+  @Override
+  public Map<String, ServerInstance> getEnabledServerInstanceMap() {
+    return new HashMap<>(_enabledServerInstanceMap);
```

Comment:
Can we add a comment to explain why copy for readers (or in future whoever changes the code) to be aware of the gotchas ?

Reply:
+1. yeah.

Reply:
Don't make a copy as it can be expensive, just return as is since we can control the caller code

Reply:
we have to make a copy since if one is looping over the entry during which the content of the map is changed due to helix update. we will not be able to guarantee correctness.